pkg/smokescreen/conntrack/conn_tracker.go

```diff
@@ -15,19 +19,111 @@ type Tracker struct {
 	Wg           *sync.WaitGroup
 	statsc       statsd.ClientInterface
 
+	SuccessRateTracker *ConnSuccessRateTracker
+
 	// A connection is idle if it has been inactive (no bytes in/out) for this
 	// many seconds.
 	IdleTimeout time.Duration
 }
 
-func NewTracker(idle time.Duration, statsc statsd.ClientInterface, logger *logrus.Logger, sd atomic.Value) *Tracker {
+// ConnSuccessRateTracker tracks statistics about the overall success rate
+// of connection attempts over some time interval (which is set in StartNewConnSuccessRateTracker()).
+//
+// It tracks only the *most recently seen* connection to an individual destination host within the configured
+// time interval, to prevent a single destination host from having an outsized
+// impact on statistics.
+type ConnSuccessRateTracker struct {
+	ConnAttempts         *cache.Cache
+	ConnSuccessRateStats atomic.Value
+}
+
+// ConnSuccessRateStats represents a timestamped output of computations performed over
+// connection attempts.
+//
+// ConnSuccessRate represents the proportion of unique domains whose most recent connection attempt
+// was successful, compared to the total number of domains with connection attempts.
+type ConnSuccessRateStats struct {
+	CalculatedAt    time.Time
+	ConnSuccessRate float64
+	TotalConns      int
+}
+
+// StartNewConnSuccessRateTracker creates a new ConnSuccessRateTracker with a specific calculation interval at which
+// ConnSuccessRateStats will be recomputed, and a time window to calculate those statistics over.
+// - calculationInterval is how often statistics will be recomputed, in seconds.
+// - calculationWindow is the period that statistics will be calculated over, in seconds.
+// - cleanupInterval is how often expired items (e.g., items older than the calculationWindow) will be deleted from memory, in seconds.
```

Comment:
```suggestion
// - calculationInterval is how often statistics will be recomputed.
// - calculationWindow is the period that statistics will be calculated over.
// - cleanupInterval is how often expired items (e.g., items older than the calculationWindow) will be deleted from memory.
```

"in seconds" would be necessary if these values weren't already `time.Duration`, like if they were `int`.

Reply:
of course! thanks :)